### ISSUE TITLE
Bigtable Emulator: Wrong row returned for read with concurrent writes at high load.

### ISSUE THREAD

**Issue by MitchRudominer** [GoogleCloudPlatform/google-cloud-go#826]
I am using btemulator on Linux, via gRPC with a C++ client. One of my unit tests is flaky and I believe I have traced it to btemulator returning the wrong row in a single-row read. It only happens sometimes and it only happens in one particular test in which there are many concurrent reads and writes.

I do a read with with a row range which is a closed interval of length one--that is it looks like:
["my-row-key", "my-row-key"] --- and I get back a single row but it is not the row with key
"my-row-key". Instead it is a different row with a very close but different row key.

Here is a slightly simplified version of my C++ code:

```c++
ReadRowsRequest req;
req.set_table_name("MyTable");
RowSet* rowset = req.mutable_rows();
RowRange* row_range = rowset->add_row_ranges();
row_range->mutable_start_key_closed()->swap(row_key);
row_range->mutable_end_key_closed()->swap(row_key); // notice end=start

ClientContext context;
std::unique_ptr<ClientReader<ReadRowsResponse>> reader(
      stub_->ReadRows(&context, req));

ReadRowsResponse resp;
 std::string current_column_name = "";
 
 while (reader->Read(&resp)) {
    for (const auto& chunk : resp.chunks()) {
    
      // When we get a different row key, start a new row.
      if (rows.empty() ||
          (!chunk.row_key().empty() &&
           rows.back().key != chunk.row_key())) {
        rows.emplace_back();
        rows.back().key = chunk.row_key();
        // We are starting a new row so reset the current column.
        current_column_name = "";
      }
      auto& row = rows.back();
      if (!chunk.has_qualifier()) {
        // Keep using the same current column.
        CHECK(!current_column_name.empty());
      } else {
        // Update the current column.
        current_column_name = chunk.qualifier().value();
      }
      row.column_values[current_column_name] += chunk.value();
      if (chunk.commit_row()) {
        num_complete_rows_read++;
      }
      CHECK(!chunk.reset_row());
    }
  }

  grpc::Status status = reader->Finish();

  if (!status.ok()) {
    LOG(ERROR) << ErrorMessage(status, "ReadRows");
  }

 DCHECK(rows.size() == 1);
 DCHECK(rows[0].key == row_key)  << rows[0].key << " vs " << row_key;
```

The final DCHECK fails sometimes with a message such as the following:

Check failed: rows[0].key == row->key 0000000001:0000000001:0000000003:00000000000001234567:2732448795:0000 vs 0000000001:0000000001:0000000003:00000000000001234567:2931898344:0000

So the read does return a single row but the row key for the returned row is different than the
row key I requested (in this case it is lexicographaically less.)

This phenomenon only happens when there are multiple threads concurrently writing to and
reading from the same range of rows. My unit test fails about one out of five times so this must
be a race condition.

I'm trying to figure out how to tell you which version of btemulator I am using. It is one that came with a recent download of the cloud SDK. If I type "gcloud --version" I see
/gcloud --version
Google Cloud SDK 180.0.1
bigtable 
bq 2.0.27
core 2017.11.20
gsutil 4.28
kubectl

**Comment by MitchRudominer** [comment]
@garye  I wanted to call your attention to
https://fuchsia-review.googlesource.com/#/c/cobalt/+/103137/

I am disabling the flaky test that I pointed you to (in a separate communication) as a means of reproducing this bug. However you can still reproduce it using the command that I mentioned, namely by passing the flag `--tests=btemulator`

**Comment by garye** [comment]
Yes, I can reproduce it easily using the command that you gave me. That's going to help a lot, thanks!

**Comment by garye** [comment]
Fixed by d594007
